rust/worker/src/server.rs

```diff
@@ -588,4 +603,187 @@ mod tests {
         assert!(response.is_err());
         assert_eq!(response.unwrap_err().code(), tonic::Code::InvalidArgument);
     }
+
+    #[tokio::test]
+    async fn validate_knn_plan() {
+        let gen_request = || chroma_proto::KnnPlan {
+            scan: Some(scan()),
+            filter: Some(chroma_proto::FilterOperator {
+                ids: None,
+                r#where: None,
+                where_document: None,
+            }),
+            knn: Some(chroma_proto::KnnOperator {
+                embeddings: vec![],
+                fetch: 0,
+            }),
+            projection: Some(chroma_proto::KnnProjectionOperator {
+                projection: Some(chroma_proto::ProjectionOperator {
+                    document: false,
+                    embedding: false,
+                    metadata: false,
+                }),
+                distance: false,
+            }),
+        };
+
+        let mut executor = QueryExecutorClient::connect(run_server()).await.unwrap();
+
+        // empty embeddings
+        let response = executor.knn(gen_request()).await;
+        assert!(response.is_ok());
+        assert_eq!(response.unwrap().into_inner().results.len(), 0);
+
+        // invalid collection
+        let mut request = gen_request();
+        request.scan = None;
+        let response = executor.knn(request).await;
+        let err = response.unwrap_err();
+        assert_eq!(err.code(), tonic::Code::InvalidArgument);
+        assert!(
+            err.message().to_lowercase().contains("scan operator"),
+            "{}",
+            err.message()
+        );
```

Comment:
I'd make each of these blocks a separate test.

Reply:
Thank you @rescrv  for the feedback! I understand that the test function remains lengthy, even after splitting into two parts (`test_knn_plan`, `test_knn_plan_scan`). The currently structure groups related sub-tests for a single function, with comments to clarify their purpose. Splitting these further would likely result in a large number of tests for the same function and introduce significant code duplication. However, I agree that refactoring could improve readability, and I open to addressing this if needed. Please let me know if you consider this a blocker, and I'II prioritise splitting them further.

Reply:
Could you make helper utilities that you call?  It seems like gen_request is ripe for that.  I may be missing why more than that is necessary to break these tests apart.  They don't seem to build on each other.

Reply:
Thanks for the feedback! I've restructured the tests as you suggested, breaking each block into its own test function. ✅ The tests are now more modular, with 9 focused tests function instead of 2 larger ones. Please let me know if this approach aligns with your expectations. 🙂  